main/lsp/updates.cc

```diff
@@ -219,58 +282,51 @@ bool LSPLoop::canTakeFastPath(const vector<shared_ptr<core::File>> &changedFiles
+        logger->debug("Trying to see if happy path is available after {} file changes", updates.updatedFiles.size());
+        ENFORCE(updates.updatedFiles.size() == hashes.size());
+        takeFastPath = canTakeFastPath(updates, hashes);
 
         int i = -1;
-        {
-            core::UnfreezeFileTable fileTableAccess(*initialGS);
-            for (auto &f : changedFiles) {
-                ++i;
-                auto fref = updateFile(f);
-                if (globalStateHashes.size() <= fref.id()) {
-                    // New file
-                    ENFORCE(!takeFastPath);
-                    globalStateHashes.resize(fref.id() + 1);
-                } else if (takeFastPath) {
-                    // Existing file on fast path
-                    auto &oldHash = globalStateHashes[fref.id()];
-                    for (auto &p : hashes[i].definitions.methodHashes) {
-                        auto fnd = oldHash.definitions.methodHashes.find(p.first);
-                        ENFORCE(fnd != oldHash.definitions.methodHashes.end(), "definitionHash should have failed");
-                        if (fnd->second != p.second) {
-                            changedHashes.emplace_back(p.first);
-                        }
+        for (auto &f : updates.updatedFiles) {
+            ++i;
+            auto fref = initialGS->findFileByPath(f->path());
+            if (fref.exists() && takeFastPath) {
+                // Update to existing file on fast path
+                auto &oldHash = globalStateHashes[fref.id()];
+                for (auto &p : hashes[i].definitions.methodHashes) {
+                    auto fnd = oldHash.definitions.methodHashes.find(p.first);
+                    ENFORCE(fnd != oldHash.definitions.methodHashes.end(), "definitionHash should have failed");
+                    if (fnd->second != p.second) {
+                        changedHashes.emplace_back(p.first);
                     }
-                    finalGs = core::GlobalState::replaceFile(move(finalGs), fref, changedFiles[i]);
-                    subset.emplace_back(fref);
                 }
-                globalStateHashes[fref.id()] = hashes[i];
+                finalGs = core::GlobalState::replaceFile(move(finalGs), fref, f);
+                subset.emplace_back(fref);
             }
-            core::NameHash::sortAndDedupe(changedHashes);
+            // Note: We may not have an id yet for this file if it is brand new, so we store hashes with their paths.
+            updates.updatedFileHashes.push_back(make_pair(f->path(), hashes[i]));
         }
+        core::NameHash::sortAndDedupe(changedHashes);
     }
 
     if (takeFastPath) {
         Timer timeit(logger, "fast_path");
-        // Drop any indexing errors produced during `updateFile`, as we are going to run indexing a second time with
-        // finalGS. (Note: Flushing is disabled in LSP mode, so we have to drain.)
-        errorQueue->drainWithQueryResponses();
```

Comment:
No longer needed now that we run `updateFile` on `initialGS` at the *end* of the typechecking run -- in `commitTypecheckRun`.